src/types/function/mod.rs

```diff
@@ -0,0 +1,140 @@
+//! Types and traits for working with JavaScript functions.
+
+use crate::handle::Handle;
+use crate::types::Value;
+
+use smallvec::smallvec;
+
+pub(crate) mod private;
+
+/// The trait for specifying arguments for a function call. This trait is sealed and cannot
+/// be implemented by types outside of the Neon crate.
+///
+/// **Note:** This trait is implemented for tuples of up to 32 JavaScript values,
+/// but for the sake of brevity, only tuples up to size 8 are shown in this documentation.
+pub trait Arguments<'a>: private::ArgumentsInternal<'a> {}
+
+impl<'a, T: Value> private::ArgumentsInternal<'a> for Vec<Handle<'a, T>> {
+    fn into_args_vec(self) -> private::ArgsVec<'a> {
+        let mut args = smallvec![];
+        for arg in self {
+            args.push(arg.upcast());
+        }
+        args
+    }
+}
+
+impl<'a, T: Value> Arguments<'a> for Vec<Handle<'a, T>> {}
+
+impl<'a, T: Value, const N: usize> private::ArgumentsInternal<'a> for [Handle<'a, T>; N] {
```

Comment:
`const` generics will require a fairly recent version of Rust (1.51, released in March). Given that we're copying anyway, maybe this only needs to be implemented over slices?

I think this and `Vec` could be deleted and only leave an implementation for `&[Handle<'a, T>]`.